File: backend/integration/efchat.go
// Copyright (C) 2025 efchat.net <tj@efchat.net>
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

package integration

import (
	"database/sql"
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"strings"
	
	"github.com/gorilla/mux"
	"github.com/redis/go-redis/v9"
	
	"github.com/efchatnet/efsec/backend/handlers"
	"github.com/efchatnet/efsec/backend/middleware"
	"github.com/efchatnet/efsec/backend/storage/postgres"
)

// E2EIntegration provides E2E encryption functionality as a plugin for efchat
type E2EIntegration struct {
	store         *postgres.Store
	keyHandler    *handlers.KeyHandler
	groupHandler  *handlers.GroupHandler
	dmHandler     *handlers.DMHandler
	spaceHandler  *handlers.SpaceHandler
	jwtSecret     string
	jwtIssuer     string
}

// Config holds configuration for the E2E integration
type Config struct {
	DB        *sql.DB
	Redis     *redis.Client
	JWTSecret string
	JWTIssuer string
}

// NewE2EIntegration creates a new E2E integration that can be embedded into efchat
func NewE2EIntegration(config *Config) (*E2EIntegration, error) {
	store := postgres.NewStore(config.DB, config.Redis)
	
	// Run migrations
	if err := store.Migrate(); err != nil {
		return nil, err
	}
	
	return &E2EIntegration{
		store:        store,
		keyHandler:   handlers.NewKeyHandler(store),
		groupHandler: handlers.NewGroupHandler(store),
		dmHandler:    handlers.NewDMHandler(store),
		spaceHandler: handlers.NewSpaceHandler(store),
		jwtSecret:    config.JWTSecret,
		jwtIssuer:    config.JWTIssuer,
	}, nil
}

// RegisterRoutes adds E2E routes to an existing router
// If authMiddleware is nil, it will use the built-in JWT validation
func (e *E2EIntegration) RegisterRoutes(router *mux.Router, authMiddleware func(http.Handler) http.Handler) {
	// Create subrouter for E2E endpoints
	api := router.PathPrefix("/api/e2e").Subrouter()
	
	// Use provided auth middleware or create our own
	if authMiddleware != nil {
		api.Use(authMiddleware)
	} else {
		// Use our own JWT validation
		api.Use(middleware.NewAuthMiddleware(e.jwtSecret, e.jwtIssuer))
	}
	
	// Key management endpoints
	api.HandleFunc("/keys", e.keyHandler.RegisterKeys).Methods("POST", "OPTIONS")
	api.HandleFunc("/bundle/{user_id}", e.keyHandler.GetPreKeyBundle).Methods("GET", "OPTIONS")
	api.HandleFunc("/keys/replenish", e.keyHandler.ReplenishPreKeys).Methods("POST", "OPTIONS")
	api.HandleFunc("/keys/status", e.GetKeyStatus).Methods("GET", "OPTIONS")
	
	// DM space endpoints
	api.HandleFunc("/dm/initiate", e.spaceHandler.InitiateDM).Methods("POST", "OPTIONS")
	api.HandleFunc("/dm/find", e.spaceHandler.FindDM).Methods("GET", "OPTIONS")
	api.HandleFunc("/dm/list", e.spaceHandler.ListDMs).Methods("GET", "OPTIONS")
	
	// Space management endpoints
	api.HandleFunc("/space/{spaceId}/type", e.spaceHandler.GetSpaceType).Methods("GET", "OPTIONS")
	api.HandleFunc("/space/{spaceId}/enable-e2e", e.spaceHandler.EnableE2EForSpace).Methods("POST", "OPTIONS")
	
	// Group endpoints
	api.HandleFunc("/group/create", e.groupHandler.CreateGroup).Methods("POST", "OPTIONS")
	api.HandleFunc("/group/create-space", e.spaceHandler.CreateE2EGroup).Methods("POST", "OPTIONS")
	api.HandleFunc("/group/{groupId}/join", e.groupHandler.JoinGroup).Methods("POST", "OPTIONS")
	api.HandleFunc("/group/{groupId}/members", e.groupHandler.GetGroupMembers).Methods("GET", "OPTIONS")
	api.HandleFunc("/group/{groupId}/message", e.groupHandler.SendGroupMessage).Methods("POST", "OPTIONS")
	api.HandleFunc("/group/{groupId}/leave", e.groupHandler.LeaveGroup).Methods("POST", "OPTIONS")
	api.HandleFunc("/group/{groupId}/rekey", e.rekeyGroup).Methods("POST", "OPTIONS")
	
	// DM endpoints (for encrypted direct messages and key distribution)
	api.HandleFunc("/dm/send", e.dmHandler.SendDM).Methods("POST", "OPTIONS")
	api.HandleFunc("/dm/messages", e.dmHandler.GetDMs).Methods("GET", "OPTIONS")
	api.HandleFunc("/dm/messages/{user_id}", e.dmHandler.GetDMsWith).Methods("GET", "OPTIONS")
	api.HandleFunc("/dm/message/{messageId}/read", e.dmHandler.MarkDMRead).Methods("POST", "OPTIONS")
	api.HandleFunc("/dm/message/{messageId}", e.dmHandler.DeleteDM).Methods("DELETE", "OPTIONS")
}

// GetStore returns the underlying storage implementation
func (e *E2EIntegration) GetStore() *postgres.Store {
	return e.store
}

// CheckPreKeyCount checks if a user needs to replenish their one-time prekeys
func (e *E2EIntegration) CheckPreKeyCount(userID string, threshold int) (bool, error) {
	count, err := e.store.GetUnusedPreKeyCount(userID)
	if err != nil {
		return false, err
	}
	return count < threshold, nil
}

// GetKeyStatus returns the current key status for a user
func (e *E2EIntegration) GetKeyStatus(w http.ResponseWriter, r *http.Request) {
	userID, ok := middleware.GetUserID(r)
	if !ok {
		http.Error(w, "Unauthorized", http.StatusUnauthorized)
		return
	}

	count, err := e.store.GetUnusedPreKeyCount(userID)
	if err != nil {
		http.Error(w, "Failed to get key count", http.StatusInternalServerError)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(map[string]interface{}{
		"remaining_keys": count,
	})
}


// rekeyGroup initiates a key rotation for a group
func (e *E2EIntegration) rekeyGroup(w http.ResponseWriter, r *http.Request) {
	userID, ok := middleware.GetUserID(r)
	if !ok {
		http.Error(w, "Unauthorized", http.StatusUnauthorized)
		return
	}

	vars := mux.Vars(r)
	groupID := vars["groupId"]

	// Check if user is a member
	members, err := e.store.GetGroupMembers(groupID)
	if err != nil {
		http.Error(w, "Group not found", http.StatusNotFound)
		return
	}

	isMember := false
	for _, member := range members {
		if member == userID {
			isMember = true
			break
		}
	}

	if !isMember {
		http.Error(w, "Not a group member", http.StatusForbidden)
		return
	}

	// Increment key version for the group
	if err := e.store.IncrementKeyVersion(groupID); err != nil {
		http.Error(w, "Failed to rotate keys", http.StatusInternalServerError)
		return
	}

	w.WriteHeader(http.StatusOK)
	w.Write([]byte(`{"status":"rekeyed"}`))
}

// ValidateSetup checks if the E2E module is properly configured
func (e *E2EIntegration) ValidateSetup() error {
	// Check database connection
	if err := e.store.Migrate(); err != nil {
		return err
	}

	// Validate JWT configuration
	if e.jwtSecret == "" {
		return &ValidationError{Message: "JWT secret is not configured"}
	}

	return nil
}

// ValidationError represents a configuration validation error
type ValidationError struct {
	Message string
}

func (e *ValidationError) Error() string {
	return e.Message
}

// Handler getters for bridge integration
func (e *E2EIntegration) GetKeyHandler() *handlers.KeyHandler {
	return e.keyHandler
}

func (e *E2EIntegration) GetDMHandler() *handlers.DMHandler {
	return e.dmHandler
}

func (e *E2EIntegration) GetGroupHandler() *handlers.GroupHandler {
	return e.groupHandler
}

func (e *E2EIntegration) GetSpaceHandler() *handlers.SpaceHandler {
	return e.spaceHandler
}


func (e *E2EIntegration) RekeyGroup(w http.ResponseWriter, r *http.Request) {
	e.rekeyGroup(w, r)
}

// CleanupDMSession removes all E2E session data for a DM space
func (e *E2EIntegration) CleanupDMSession(spaceID string) error {
	// Extract user IDs from DM space ID format: dm-user1-user2
	if !strings.HasPrefix(spaceID, "dm-") {
		return nil // Not a DM space
	}
	
	// Parse DM space ID to get the two user IDs
	parts := strings.Split(spaceID, "-")
	if len(parts) < 3 {
		return fmt.Errorf("invalid DM space ID format: %s", spaceID)
	}
	
	user1 := parts[1]
	user2 := parts[2]
	
	// Clean up DM records between these users
	if err := e.store.DeleteDMsBetweenUsers(user1, user2); err != nil {
		return fmt.Errorf("failed to delete DMs between %s and %s: %w", user1, user2, err)
	}
	
	log.Printf("[E2E-Cleanup] Cleaned up DM session data for space %s (%s <-> %s)", spaceID, user1, user2)
	return nil
}

File: client/src/integration/efchat.ts
// Copyright (C) 2025 efchat.net <tj@efchat.net>
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

/**
 * Graceful E2E Integration for efchat
 *
 * This module provides a fail-safe integration layer that allows efchat
 * to continue working normally even if E2E encryption fails or is unavailable.
 */

import { EfSecClient } from '../index';

export interface E2EIntegrationConfig {
  apiUrl?: string;
  getAuthToken: () => string | null;
  onE2EStatusChange?: (_enabled: boolean) => void;
  debug?: boolean;
}

export interface MessageEnvelope {
  content: string;
  encrypted?: boolean;
  encryptionData?: {
    type: string;
    deviceId?: number;
  };
}

export interface WebSocketMessage {
  type: string;
  content?: string;
  isGroup?: boolean;
  encrypted?: boolean;
  encryptionData?: {
    type: string;
    deviceId?: number;
  };
  wasEncrypted?: boolean;
}

/**
 * Safe E2E wrapper that gracefully falls back to unencrypted messaging
 */
export class E2EIntegration {
  private client?: EfSecClient;
  private isAvailable: boolean = false;
  private config: E2EIntegrationConfig;
  private initPromise?: Promise<void>;

  constructor(config: E2EIntegrationConfig) {
    this.config = config;
    this.checkAvailability();
  }

  /**
   * Check if E2E module is available
   */
  private async checkAvailability(): Promise<void> {
    try {
      // EfSecClient should always be available since it's our own implementation
      this.isAvailable = true;
      if (this.config.debug) {
        console.error('E2E encryption module loaded successfully');
      }
    } catch (error) {
      this.isAvailable = false;
      console.error('E2E encryption module not available, continuing without E2E', error);

      if (this.config.onE2EStatusChange) {
        this.config.onE2EStatusChange(false);
      }
    }
  }

  /**
   * Initialize E2E if available (non-blocking)
   */
  async initializeE2E(userId: string): Promise<boolean> {
    if (!this.isAvailable) {
      return false;
    }

    // Prevent multiple initializations
    if (this.initPromise) {
      await this.initPromise;
      return this.client !== undefined;
    }

    this.initPromise = this.performInitialization(userId);

    try {
      await this.initPromise;
      return true;
    } catch (error) {
      console.error('E2E initialization failed, continuing without encryption', error);
      return false;
    }
  }

  private async performInitialization(userId: string): Promise<void> {
    const token = this.config.getAuthToken();
    if (!token) {
      throw new Error('No auth token for E2E');
    }

    // Initialize EfSec client
    this.client = new EfSecClient(this.config.apiUrl ?? '/api/e2e');
    await this.client.init(token);

    if (this.config.onE2EStatusChange) {
      this.config.onE2EStatusChange(true);
    }

    if (this.config.debug) {
      console.error('E2E encryption initialized for user:', userId);
    }
  }

  /**
   * Encrypt message if E2E is available, otherwise return plaintext
   */
  async encryptMessage(
    recipientId: string,
    message: string,
    isGroup: boolean = false
  ): Promise<MessageEnvelope> {
    if (!this.client) {
      // E2E not available, send unencrypted
      return {
        content: message,
        encrypted: false,
      };
    }

    try {
      if (isGroup) {
        const encrypted = await this.client.encryptGroupMessage(recipientId, message);
        return {
          content: btoa(String.fromCharCode(...encrypted)),
          encrypted: true,
          encryptionData: { type: 'group' },
        };
      } else {
        const encrypted = await this.client.encryptDM(recipientId, message);
        return {
          content: btoa(String.fromCharCode(...encrypted)),
          encrypted: true,
          encryptionData: { type: 'dm' },
        };
      }
    } catch (error) {
      console.error('Encryption failed, sending unencrypted:', error);
      return {
        content: message,
        encrypted: false,
      };
    }
  }

  /**
   * Decrypt message if E2E is available and message is encrypted
   */
  async decryptMessage(
    senderId: string,
    envelope: MessageEnvelope,
    isGroup: boolean = false
  ): Promise<string> {
    // If not encrypted or E2E not available, return as-is
    if (!envelope.encrypted || !this.client) {
      return envelope.content;
    }

    try {
      // Convert base64 back to Uint8Array
      const binaryString = atob(envelope.content);
      const bytes = new Uint8Array(binaryString.length);
      for (let i = 0; i < binaryString.length; i++) {
        bytes[i] = binaryString.charCodeAt(i);
      }

      if (isGroup && envelope.encryptionData?.type === 'group') {
        return await this.client.decryptGroupMessage(
          senderId, // Assuming this is groupId
          senderId,
          1, // device ID - Matrix Protocol default device ID
          bytes
        );
      } else {
        return await this.client.decryptDM(senderId, bytes);
      }
    } catch (error) {
      console.error('Decryption failed, showing encrypted content:', error);
      return envelope.content;
    }
  }

  /**
   * Initiate DM if E2E is available
   */
  async initiateDM(peerId: string): Promise<string | null> {
    if (!this.client) {
      console.error('E2E client not available');
      return null;
    }

    try {
      await this.client.startDMSession(peerId);
      return `dm_${peerId}`;
    } catch (error) {
      console.error('Failed to initiate E2E DM:', error);
      return null;
    }
  }

  /**
   * Check if E2E is ready
   */
  isE2EReady(): boolean {
    return this.isAvailable && this.client !== undefined;
  }

  /**
   * Get E2E status
   */
  getStatus(): {
    available: boolean;
    initialized: boolean;
    hasSession: boolean;
  } {
    return {
      available: this.isAvailable,
      initialized: this.client !== undefined,
      hasSession: false, // Would need to check specific session
    };
  }

  /**
   * Cleanup E2E resources
   */
  async cleanup(): Promise<void> {
    if (this.client) {
      delete this.client;
    }
  }
}

/**
 * WebSocket message interceptor for automatic E2E handling
 */
export class E2EWebSocketInterceptor {
  private e2e: E2EIntegration;

  constructor(e2e: E2EIntegration) {
    this.e2e = e2e;
  }

  /**
   * Process outgoing message
   */
  async processOutgoing(
    message: WebSocketMessage,
    recipientId?: string
  ): Promise<WebSocketMessage> {
    // Only process text messages
    if (message.type !== 'message' || !message.content || !recipientId) {
      return message;
    }

    const envelope = await this.e2e.encryptMessage(recipientId, message.content, message.isGroup);

    const result: WebSocketMessage = {
      ...message,
      content: envelope.content,
    };

    if (envelope.encrypted !== undefined) {
      result.encrypted = envelope.encrypted;
    }

    if (envelope.encryptionData !== undefined) {
      result.encryptionData = envelope.encryptionData;
    }

    return result;
  }

  /**
   * Process incoming message
   */
  async processIncoming(message: WebSocketMessage, senderId?: string): Promise<WebSocketMessage> {
    // Only process encrypted messages
    if (!message.encrypted || !message.content || !senderId) {
      return message;
    }

    const envelope: MessageEnvelope = {
      content: message.content,
      encrypted: message.encrypted,
    };

    if (message.encryptionData !== undefined) {
      envelope.encryptionData = message.encryptionData;
    }

    const decrypted = await this.e2e.decryptMessage(senderId, envelope, message.isGroup);

    return {
      ...message,
      content: decrypted,
      wasEncrypted: true,
    };
  }
}

/**
 * Factory function for safe E2E integration
 */
export function createE2EIntegration(config: E2EIntegrationConfig): E2EIntegration {
  return new E2EIntegration(config);
}
